### PR TITLE
Added argocd pkg

### PR DESCRIPTION
**Pull request by shaior**
added argocd package

### CODE REVIEW

File: pkg/argocd/applications.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	argocd "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ApplicationBuilder struct {
```
Comment (kononovn) [inline pkg/argocd/applications.go:16]: Please add comments to all public elements

File: pkg/argocd/applications.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	Definition *argocd.Application
+	Object     *argocd.Application
+	apiClient  *clients.Settings
+	errorMsg   string
+}
+
+// Pull pulls existing ...
```
Comment (kononovn) [inline pkg/argocd/applications.go:23]: Please fix comment

File: pkg/argocd/applications.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return argocd, err
+}
+
+// validate will check that the builder and builder definition are properly initialized before
+// accessing any member fields.
+func (builder *ApplicationBuilder) validate() (bool, error) {
```
Comment (kononovn) [inline pkg/argocd/applications.go:97]: Please move private method to the end of file. Better to keep all public methods on top of file

File: pkg/argocd/argocd.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Builder provides struct for the argocd object containing connection to
+// the cluster and the argocd definitions.
+type Builder struct {
+	Definition *argocdoperatorv1alpha1.ArgoCD
```
Comment (kononovn) [inline pkg/argocd/argocd.go:19]: Please add comments to public elements

File: pkg/argocd/argocd.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	Object     *argocdoperatorv1alpha1.ArgoCD
+	apiClient  *clients.Settings
+	errorMsg   string
+}
+
+// AdditionalOptions additional options for argocd object.
+type AdditionalOptions func(builder *Builder) (*Builder, error)
+
+// NewBuilder creates a new instance of Builder.
+func NewBuilder(
```
Comment (kononovn) [inline pkg/argocd/argocd.go:29]: All arguments can fit one line.
func NewBuilder(apiClient *clients.Settings, name,nsname string) *Builder

File: pkg/argocd/argocd.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return argocd, err
+}
+
+// validate will check that the builder and builder definition are properly initialized before
+// accessing any member fields.
+func (builder *Builder) validate() (bool, error) {
```
Comment (kononovn) [inline pkg/argocd/argocd.go:133]: Please move private method to the end of the file.

**Review by kononovn — CHANGES_REQUESTED** [review]
Looks good, just few minor comments to fix.

File: pkg/argocd/argocd.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+	// created argocd object.
+	Object *argocdoperatorv1alpha1.ArgoCD
+	// api client to interact with the cluster.
+	apiClient *clients.Settings
+	// used to store latest error message upon defining the argocd definition.
+	errorMsg string
+}
+
+// AdditionalOptions additional options for argocd object.
+type AdditionalOptions func(builder *Builder) (*Builder, error)
```
Comment (kononovn) [inline pkg/argocd/argocd.go:30]: Please remove it. You are not using it in mutation functions.

**Review by kononovn — CHANGES_REQUESTED** [review]
1 more comment to fix

File: pkg/argocd/argocd.go (hunk 7 above)
Comment (shaior) [reply pkg/argocd/argocd.go]: Done.

**Review by kononovn — DISMISSED** [review]
lgtm

File: pkg/argocd/applications.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	return argocd, err
+}
+
+// Update renovates the existing argocd application object with the argocd application definition in builder.
+func (builder *ApplicationBuilder) Update(force bool) (*ApplicationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Updating the argocd application object", builder.Definition.Name)
```
Comment (trewest) [inline pkg/argocd/applications.go:106]:
```suggestion
	glog.V(100).Infof("Updating the argocd application object %s in namespace", builder.Definition.Name, builder.Definition.Namespace)
```

File: pkg/argocd/applications.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	return builder, nil
+}
+
+// Create makes an argocd application in the cluster and stores the created object in a struct.
+func (builder *ApplicationBuilder) Create() (*ApplicationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating argocd application %s", builder.Definition.Name)
```
Comment (trewest) [inline pkg/argocd/applications.go:158]: Please add namespace

File: pkg/argocd/applications.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	return builder, err
+}
+
+// Delete removes the argocd application object from a cluster.
+func (builder *ApplicationBuilder) Delete() (*ApplicationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Deleting the argocd application object %s", builder.Definition.Name)
```
Comment (trewest) [inline pkg/argocd/applications.go:139]: Please add namespace

**Review by trewest — CHANGES_REQUESTED** [review]
Few minor comments, otherwise looks good

**Review by trewest — APPROVED** [review]
/lgtm

File: pkg/argocd/applications.go [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+	Definition *argocd.Application
+	// created application object.
+	Object *argocd.Application
+	// api client to interact with the cluster.
+	apiClient *clients.Settings
+	// used to store latest error message upon defining or mutating application definition.
+	errorMsg string
+}
+
+// Pull pulls existing application into ApplicationBuilder struct.
```
Comment (achuzhoy) [inline pkg/argocd/applications.go:28]: nit: PullApplication

**Comment by trewest** [comment]
@kononovn I just noticed this PR is adding both ApplicationBuilder and Builder structs into. I think if multiple structs are in the same pkg, we should make both specific (i.e. ApplicationBuilder and ArgCDBuilder). WDYT?

File: pkg/argocd/argocd.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+	}, argocd)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return argocd, err
+}
+
+// Create makes a argocd in the cluster and stores the created object in struct.
```
Comment (achuzhoy) [inline pkg/argocd/argocd.go:129]: nit: makes an argocd

**Comment by kononovn** [comment]
> @kononovn I just noticed this PR is adding both ApplicationBuilder and Builder structs into the argocd pkg. I think if multiple structs are in the same pkg, we should make both specific (i.e. ApplicationBuilder and ArgCDBuilder). WDYT?

@trewest No need considering package name. Code usage will look like:
arogcd.NewBuilder
argocd.ApplicationBuilder
We do have argocd in package name

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**